Tests/Test.cs

```diff
@@ -0,0 +1,327 @@
+                        OpenedAt = openedAt
+                    });
+                    break;
+                case (ProductAdded(var (productId, quantity, price), var addedAt), not null):
+                    var toUpdate = current.ProductItems.SingleOrDefault(p => p.ProductId == productId);
+
+                    if (toUpdate == null)
+                    {
+                        current.ProductItems.Add(new ShoppingCartModel.PricedProductItem
+                        {
+                            ProductId = productId, Quantity = quantity, Price = price, AddedAt = addedAt,
+                        });
+                    }
+                    else
+                    {
+                        toUpdate.Quantity += quantity;
+                        toUpdate.UpdatedAt = addedAt;
+                    }
+
+                    break;
+                case (ProductRemoved(var (productId, quantity, price), var removedAt), not null):
+                    var existing = current.ProductItems.Single(p => p.ProductId == productId);
+
+                    if (existing.Quantity - quantity == 0)
+                    {
+                        current.ProductItems.Remove(existing);
+                    }
+                    else
+                    {
+                        existing.Quantity -= quantity;
+                        existing.UpdatedAt = removedAt;
+                    }
+
+                    break;
+                case (Confirmed (var clientId, var confirmedAt), not null):
+                    current.Status = ShoppingCartModel.ShoppingCartStatus.Confirmed;
+                    current.ClientId = clientId;
+                    current.ConfirmedAt = confirmedAt;
+                    dbContext.ShoppingCarts.Update(current);
+                    break;
+                case (Cancelled, not null):
+                    dbContext.ShoppingCarts.Remove(current);
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(@event));
+            }
+        }
+
+        return dbContext.SaveChangesAsync(ct);
+    }
+}
+
+public interface IProductPriceCalculator
+{
+    IReadOnlyList<PricedProductItem> Calculate(params ProductItem[] productItems);
+}
+
+public class DummyProductPriceCalculator(decimal price): IProductPriceCalculator
+{
+    public IReadOnlyList<PricedProductItem> Calculate(params ProductItem[] productItems) =>
+        productItems.Select(pi => new PricedProductItem(pi.ProductId, pi.Quantity, price)).ToArray();
+}
+
+public class Test
+{
+    [Fact]
+    public void GivenEmptyShoppingCart_WhenAddProduct_ThenReturnsOpenedAndProductAddedEvents()
+    {
+        //Given
+        var price = (decimal)new Random().NextDouble() * 100m;
+        var now = DateTimeOffset.Now;
+
+        var cardId = Guid.NewGuid();
+        var productItem = new ProductItem(Guid.NewGuid(), 10);
+
+        var state = new Empty();
```

Comment:
Hm not sure having the `using` make this name be available unqualified helps legibility
i.e. I would consider renaming
`public abstract record ShoppingCart` to `public abstract record State`, then you have State.*

And probably the same for ShoppingCartCommand (rename to Command, use as a prefix)

Reply:
Yes, naming is hard. But for me, `State` is too vague and generic.

Reply:
the point is that mingling the states, commands and events at top level is only a good plan in very small toys. Even in a medium sized model, it's better to have a plan to let people bucket human sized bites of things